docs/source/amp.rst

```diff
@@ -256,6 +256,141 @@ In this case, combine the two layers using :func:`torch.nn.functional.binary_cro
+please file an issue.
+
+XPU Ops that can autocast to ``bfloat16``
+""""""""""""""""""""""""""""""""""""""""
+
+``__matmul__``,
+``addbmm``,
+``addmm``,
+``addmv``,
+``addr``,
+``baddbmm``,
+``bmm``,
+``chain_matmul``,
+``multi_dot``,
+``conv1d``,
+``conv2d``,
+``conv3d``,
+``conv_transpose1d``,
+``conv_transpose2d``,
+``conv_transpose3d``,
+``GRUCell``,
+``linear``,
+``LSTMCell``,
+``matmul``,
+``mm``,
+``mv``,
+``prelu``,
+``RNNCell``
+
+XPU Ops that can autocast to ``float32``
+""""""""""""""""""""""""""""""""""""""""
+
+``__pow__``,
+``__rdiv__``,
+``__rpow__``,
+``__rtruediv__``,
+``acos``,
+``asin``,
+``binary_cross_entropy_with_logits``,
+``cosh``,
+``cosine_embedding_loss``,
+``cdist``,
+``cosine_similarity``,
+``cross_entropy``,
+``cumprod``,
+``cumsum``,
+``dist``,
+``erfinv``,
+``exp``,
+``expm1``,
+``group_norm``,
+``hinge_embedding_loss``,
+``kl_div``,
+``l1_loss``,
+``layer_norm``,
+``log``,
+``log_softmax``,
+``log10``,
+``log1p``,
+``log2``,
+``margin_ranking_loss``,
+``mse_loss``,
+``multilabel_margin_loss``,
+``multi_margin_loss``,
+``nll_loss``,
+``norm``,
+``normalize``,
+``pdist``,
+``poisson_nll_loss``,
+``pow``,
+``prod``,
+``reciprocal``,
+``rsqrt``,
+``sinh``,
+``smooth_l1_loss``,
+``soft_margin_loss``,
+``softmax``,
+``softmin``,
+``softplus``,
+``sum``,
+``renorm``,
+``tan``,
+``triplet_margin_loss``
+
+XPU Ops that promote to the widest input type
+"""""""""""""""""""""""""""""""""""""""""""""
+These ops don't require a particular dtype for stability, but take multiple inputs
+and require that the inputs' dtypes match.  If all of the inputs are
+``bfloat16``, the op runs in ``bfloat16``.  If any of the inputs is ``float32``,
+autocast casts all inputs to ``float32`` and runs the op in ``float32``.
+
+``addcdiv``,
+``addcmul``,
+``atan2``,
+``bilinear``,
+``cross``,
+``dot``,
+``grid_sample``,
+``index_put``,
+``scatter_add``,
+``tensordot``
+
+Some ops not listed here (e.g., binary ops like ``add``) natively promote
+inputs without autocasting's intervention.  If inputs are a mixture of ``bfloat16``
+and ``float32``, these ops run in ``float32`` and produce ``float32`` output,
+regardless of whether autocast is enabled.
+
+Prefer ``binary_cross_entropy_with_logits`` over ``binary_cross_entropy``
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+The backward passes of :func:`torch.nn.functional.binary_cross_entropy` (and :mod:`torch.nn.BCELoss`, which wraps it)
+can produce gradients that aren't representable in ``bfloat16``.  In autocast-enabled regions, the forward input
+may be ``bfloat16``, which means the backward gradient must be representable in ``bfloat16`` (autocasting ``bfloat16``
+forward inputs to ``float32`` doesn't help, because that cast must be reversed in backward).
+Therefore, ``binary_cross_entropy`` and ``BCELoss`` raise an error in autocast-enabled regions.
+
+Many models use a sigmoid layer right before the binary cross entropy layer.
```

Comment:
sigmoid is currently not supported by XPU [1]. Not sure about other operations mentioned here.

[1] https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L380

Reply:
Ditto